draft-ietf-ace-oscore-profile.xml

```diff
@@ -773,7 +773,7 @@ token.
         </t>
 
         <t>
-          If the exchange was an update of access rights, i.e., a new Security Context was derived from a client that already had a Security Context in place, the RS is RECOMMENDED to delete the old Security Context after OSCORE verification and verification of access rights succeed. The RS MUST delete the Security Context if it deletes the access token associated to it.
+          If the exchange was an update of access rights, i.e., a new Security Context was derived from a client that already had a Security Context in place, the RS is RECOMMENDED to delete the old Security Context after OSCORE verification succeeds. The RS MUST delete the Security Context if it deletes the access token associated to it.
```

Comment:
I do see that the previous paragraph talks about "verification of access rights" which in some sense is parallel to the structure I'm removing here, but IIUC that paragraph is considering the generic case of "verify requests and send responses to C using OSCORE"; this paragraph seems to be specific to the "get a new token and update access rights" operation, which does not have the same type of ongoing access rights verification to perform.

Reply:
Yes, this paragraph is specific to update of access right, but I think this text should not be removed. What I want to say is that the old security context is recommended to be deleted from the RS only after successful processing of a request with that new security context, and that includes both OSCORE verification and access right verification.

Reply:
That seems like a reasonable behavior to recommend, but IIUC it seems like it would involve a period where two tokens are stored -- one "active" and one "provisional" since the opportunity to process a request with the new security context would not occur in the same transaction as the authz-info endpoint upload.  Do you want to add some more text about this?

Reply:
Actually, about 2 tokens being stored at the same time, that is covered by the ace framework and reported in this document (in section 4.2):    

> As specified in section 5.8.1 of [I-D.ietf-ace-oauth-authz], when
>    receiving an updated access token with updated authorization
>    information from the client (see Section 3.1), it is recommended that
>    the RS overwrites the previous token, that is only the latest
>    authorization information in the token received by the RS is valid.
>    This simplifies for the RS to keep track of authorization information
>    for a given client.

So no, only one token is RECOMMENDED to be kept, and that happens before the new security context is derived and "approved" (so before 2 Sec ctx exist at the same time in the RS).
Do we need some more text about this? I am not really sure what is missing... Also looking back at the open issues, I am noticing that #21 by @jimsch might be somewhat related to this (this paragraph was added to answer that issue).

Reply:
I'm trying to consider the series of events for this "updated authorization information" flow.
Presuming we start in a situation where the client has a valid token and security context with the RS, and has just obtained an updated token (and new key) from the AS. Then:

* the client posts the new token to the authz-info endpoint
* the RS validates the token (signature by AS, etc.) and notes that this token uses a different key than the current token/security context
* ???
* the RS receives an OSCORE request with the new key and is able to validate the proof of possession.  At this stage it is appropriate to delete the old token and security context (and also to validate the access rights for the OSCORE request providing the proof of possession).

I'm not sure what happens in the "???", where the RS knows that there's a new token+key but doesn't have a proof of possession for it yet.  I don't see a way for the RS to escape storing two tokens during that time without being trivially vulnerable to a DoS attack.

Reply:
I also don't know in ??? how the RS determines that it is the same client.

Reply:
@Ben, you are correct, RS would have had to keep 2 tokens (one valid, one not yet valid) for the amount of time it takes to validate the OSCORE request.
I have a proposal that I posted to the list as it is a big enough change that I wanted to get feedback there: https://mailarchive.ietf.org/arch/msg/ace/dLkW-MYHXfZqmtY7AP7ZBDJpxOw/